src/option.c

```diff
@@ -537,3 +539,83 @@ coap_option_filter_get(const coap_opt_filter_t filter, uint16_t type) {
    * but as _set and _unset do, the function does not take a const). */
   return coap_option_filter_op((uint16_t *)filter, type, FILTER_GET);
 }
+
+coap_optlist_t *coap_new_optlist(uint16_t number,
+                                 size_t length,
+                                 const uint8_t *data) {
+  coap_optlist_t *node;          
+     
+  node = coap_malloc_type(COAP_OPTLIST, sizeof(coap_optlist_t) + length);
+     
+  if (node) {
+    node->number = number;
+    node->length = length;
+    memcpy(node->data, data, length);
+  } else {
+    coap_log(LOG_DEBUG, "coap_new_optlist: malloc failure\n");
+  }
+  
+  return node;
+}
+
+static int
+order_opts(void *a, void *b) {
+  coap_optlist_t *o1 = (coap_optlist_t *)a;
+  coap_optlist_t *o2 = (coap_optlist_t *)b;
+
+  if (!a || !b)
+    return a < b ? -1 : 1;
+
+  return (int)(o1->number - o2->number);
+}
+
+int
+coap_add_optlist_pdu(coap_pdu_t *pdu, coap_optlist_t* options) {
+  coap_optlist_t *opt;
+
+  if (options) {
+    /* sort options for delta encoding */
+    LL_SORT((options), order_opts);
+
+    LL_FOREACH((options), opt) {
+      coap_add_option(pdu, opt->number, opt->length, opt->data);
+    }
+    return 1;
+  }
+  return 0;
+}
+
+int
+coap_insert_optlist(coap_optlist_t **head, coap_optlist_t *node) {
+  if (!node) {
+    assert(node);
```

Comment:
This assert makes the usage pattern `coap_insert_optlist(&list, coap_new_optlist(...))` difficult as mentioned for `client.c` where it is used heavily. In my opinion, the warning and false return value are sufficient.

Reply:
Fixed in Update - assert() is removed